modules/core/src/main/scala/com.snowplowanalytics.snowplow.bigquery/processing/TableManager.scala

```diff
@@ -44,75 +47,114 @@ object TableManager {
-            }
-      }
+    def addColumns(columns: Vector[Field]): F[TableManager.AddColumnsResult] =
+      for {
+        table <- Sync[F].blocking(client.getTable(config.dataset, config.table))
+        schema <- Sync[F].pure(table.getDefinition[TableDefinition].getSchema)
+        fields <- Sync[F].pure(schema.getFields)
+        fields <- Sync[F].pure(BigQuerySchemaUtils.mergeInColumns(fields, columns))
+        schema <- Sync[F].pure(Schema.of(fields))
+        table <- Sync[F].pure(setTableSchema(table, schema))
+        result <- Sync[F]
+                    .blocking(table.update())
+                    .as[AddColumnsResult](AddColumnsResult.Success)
+                    .recover(handleTooManyColumns)
+                    .onError(logOnRaceCondition)
+      } yield result
+
+    def createTable: F[Unit] = {
+      val tableInfo = atomicTableInfo(config)
+      Logger[F].info(show"Creating table $tableInfo") *>
+        Sync[F]
+          .blocking(client.create(tableInfo))
+          .void
+          .recoverWith {
+            case bqe: BigQueryException if bqe.lowerCaseReason === "duplicate" =>
+              // Table already exists
+              Logger[F].info(s"Ignoring error when creating table: ${bqe.getMessage}")
+            case bqe: BigQueryException if bqe.lowerCaseReason === "accessdenied" =>
+              Logger[F].info(s"Access denied when trying to create table. Will ignore error and assume table already exists.")
+          }
+    }
   }
 
-  private def addColumnsImpl[F[_]: Sync](
-    config: Config.BigQuery,
-    client: BigQuery,
-    columns: Vector[Field]
-  ): F[Unit] =
-    for {
-      table <- Sync[F].blocking(client.getTable(config.dataset, config.table))
-      schema <- Sync[F].pure(table.getDefinition[TableDefinition].getSchema)
-      fields <- Sync[F].pure(schema.getFields)
-      fields <- Sync[F].pure(BigQuerySchemaUtils.mergeInColumns(fields, columns))
-      schema <- Sync[F].pure(Schema.of(fields))
-      table <- Sync[F].pure(setTableSchema(table, schema))
-      _ <- Sync[F]
-             .blocking(table.update())
-             .void
-             .onError(logOnRaceCondition)
-    } yield ()
-
   private def setTableSchema(table: Table, schema: Schema): Table =
     table.toBuilder().setDefinition(StandardTableDefinition.of(schema)).build()
 
   private def logOnRaceCondition[F[_]: Sync]: PartialFunction[Throwable, F[Unit]] = {
-    case BigQueryUtils.BQExceptionWithLowerCaseReason("invalid") =>
+    case bqe: BigQueryException if bqe.lowerCaseReason === "invalid" =>
       Logger[F].warn(s"Caught known exception which probably means another loader has already altered the table.")
     // Don't do anything else; the BigQueryRetrying will handle retries and logging the exception.
   }
 
+  private def handleTooManyColumns: PartialFunction[Throwable, AddColumnsResult] = {
+    case bqe: BigQueryException if bqe.lowerCaseReason === "invalid" && bqe.lowerCaseMessage.startsWith("too many columns") =>
```

Comment:
Whole `TableManager` is mocked in our unit tests, so it means we're not able to unit test this part. 
On the other hand It looks like `withHandledErrors` is testable, so maybe we should move all exception/error handling logic there? 

I mean, `addColumns` and `createTable` in `TableManager` wouldn't then contain any `recover` or `onError` blocks. These would be in the new `WithHandledErrors` (that would also match its name I guess?). Then we could also throw `BigQueryException` in tests explicitly and verify reason/case/message match logic...?

Reply:
Nice idea.  I just pushed an extra commit with that suggested change.